New and improved DevTools.

grokys opened the pull request:
## What does the pull request do?

Re-implements DevTools with a few new features:

- Adds a built-in console using roslyn scripting which allows running arbitrary code
- Allow editing of property values
- Improved display and grouping of control properties
- Filtering control properties
- Maintain selection when switching between logical/visual tree

![image](https://user-images.githubusercontent.com/1775141/73022951-d1586080-3e2a-11ea-8520-84ab3319d416.png)

## Breaking Changes

`Avalonia.Diagnostics` is now a separate NuGet package so if you're using `AttachDevTools` you'll have to add a reference to that.

grokys committed:
Initial implementation of new DevTools.

kekekeks commented:
>Adds a built-in console using roslyn scripting which allows running arbitrary code

Can we have that part to be plugable? Preferably shipped in a separate package. So people won't need to ship the entire compiler with the app to use dev tools. Also not that it's not really linker-friendly and definitely not iOS-friendly

grokys commented:
> Can we have that part to be pluggable?

We could yeah, but I've got a bunch of stuff to finish that I think are higher-priority (such as `ItemsRepeater` as `ItemsPresenter`, fixing selection, fixing `ScrollViewer` etc) so I'm not sure when I'd be able to do it (this code was mostly already written, I just updated it to `master` and tidied it up a bit).

Given that, I think there are a few options:

- Keep current devtools until I can make the console pluggable
- Merge this devtools but remove the console for now
- Tell people to `#if` the call to attach the devtools if they don't want a dependency on roslyn scripting

Not sure what people would prefer?

Gillibald commented:
Apps don't rely on DevTools so it's fine if the console breaks on iOS. I am sure there will be a solution for iOS the moment iOS support is stable.

MarchingCube commented:
I think new DevTools are quite a step forward as it is much more usable interface. Console is also really handy for development. We should give an option to the people to completely remove it, maybe entire diagnostics needs to be pluggable in the end. But for now since we just released new version I think it will be fine to have this for a while until we come up with an idea for separating this feature.

kekekeks commented:
`Avalonia.Diagnostics` library is a part of `Avalonia` package. So this change effectively adds a dependency on `Microsoft.CodeAnalysis.CSharp.Scripting` to every single avalonia-based application.

Gillibald commented:
And why exactly do we need a dependency on `Avalonia.Diagnostics` in the core of Avalonia? I think that is the only issue here. This is not required to run an application.

grokys commented:
Note that `Avalonia.Diagnostics` now also has a dependency on `Avalonia.Control.DataGrid`.

Thing is, afaik there's no so thing as "Debug build-only" dependency. You either have the dependency or you don't. The best you can do it to not reference code in the dependency when you don't want to ship it. Or am I mistaken?

grokys commented:
https://github.com/NuGet/Home/issues/5895

pr8x commented:
Hi I tested your PR locally and it seems to be working fine so far. Editing `Color`, `Size` and some other types is broken however.
I would suggest adding the Type of the property (`Property.PropertyType.Name`) in the `DataGrid` as well. This will make editing a bit easier, because you don't have to manually lookup the type of the property. Also making the column headers visible and resizable would be great as well.

![image](https://user-images.githubusercontent.com/4670166/73121708-40de6500-3f7d-11ea-9581-de18794b4159.png)

MarchingCube commented:
One extra potential improvement: Allow for searching for event types in the `Events` tab. It quickly gets tedious to find wanted event type every time you open DevTools.

kekekeks commented:
>Note that Avalonia.Diagnostics now also has a dependency on Avalonia.Control.DataGrid.

We probably need to extract it to a separate package then.

grokys committed:
Merge branch 'master' into feature/new-devtools

grokys committed:
Show column headers and allow resizing.

grokys committed:
Display property types.

grokys committed:
Correctly convert strings to property value type.

grokys committed:
Remove diagnostics from core libraries.

This prevents them being added to the core `Avalonia.nupkg`. Because diagnostics now has a dependency on `DataGrid` and roslyn scripting, ship it as a separate package.

grokys committed:
Merge branch 'master' into feature/new-devtools

grokys committed:
Remove Avalonia.Dialogs dependency on diagnostics.

And add explicit dependency on `Avalonia.Diagnostics` to samples.

grokys committed:
Use TypeConverter correctly.

grokys commented:
This is now ready for review, I've made most of the suggested changes:

- Avalonia.Diagnostics is now a separate package
- Added "Type" column to show property type
- Editing colors, sizes etc now works

I didn't add the filter to the Events tab yet as that's a bit more involved and I'd like to get this finished for the moment so I can continue with other things.

MarchingCube commented on src/Avalonia.Diagnostics/DevToolsExtensions.cs:
```diff
@@ -0,0 +1,19 @@
+
+namespace Avalonia
+{
+    public static class DevToolsExtensions
```
Is it the time we start documenting public APIs ? :)

grokys replied:
It is, yes! :)

MarchingCube commented on src/Avalonia.Diagnostics/Diagnostics/Models/ConsoleHistoryItem.cs:
```diff
@@ -0,0 +1,19 @@
+
+namespace Avalonia.Diagnostics.Models
+{
+    public class ConsoleHistoryItem
```
`internal` I guess?

MarchingCube commented on src/Avalonia.Diagnostics/Diagnostics/ViewModels/ClrPropertyViewModel.cs:
```diff
@@ -0,0 +1,58 @@
+        private readonly object _target;
+        private string _type;
+        private object _value;
+        private TypeConverter _converter;
```
Unused.

MarchingCube commented on src/Avalonia.Diagnostics/Diagnostics/ViewModels/AvaloniaPropertyViewModel.cs:
```diff
@@ -0,0 +1,97 @@
+        private string _type;
+        private object _value;
+        private string _priority;
+        private TypeConverter _converter;
```
Unused.

MarchingCube commented on src/Avalonia.Diagnostics/Diagnostics/ViewModels/AvaloniaPropertyViewModel.cs:
```diff
@@ -0,0 +1,97 @@
+        private string _priority;
+        private TypeConverter _converter;
+        private string _group;
+        private DataGridCollectionView _bindingsView;
```
Unused.

MarchingCube commented on src/Avalonia.Diagnostics/Diagnostics/ViewModels/ConsoleViewModel.cs:
```diff
@@ -0,0 +1,112 @@
+{
+    internal class ConsoleViewModel : ViewModelBase
+    {
+        readonly ConsoleContext _context;
```
Nit: Missing access modifiers.

MarchingCube commented on src/Avalonia.Diagnostics/Diagnostics/Models/ConsoleContext.cs:
```diff
@@ -0,0 +1,36 @@
+
+namespace Avalonia.Diagnostics.Models
+{
+    public class ConsoleContext
```
Does this have to be `public` ?

grokys replied:
That does seem to be the case unfortunately:

![image](https://user-images.githubusercontent.com/1775141/74679014-e66fa780-51bc-11ea-9bd9-aefa63e4e35a.png)

I wonder if there's any way around this?

MarchingCube replied:
No idea, just trying to minimize public API if possible.

MarchingCube commented on src/Avalonia.Diagnostics/Diagnostics/ViewModels/ControlDetailsViewModel.cs:
```diff
@@ -0,0 +1,185 @@
+            }
+        }
+
+        public IEnumerable<string> Classes
```
Looks like to be unused.

MarchingCube commented on src/Avalonia.Diagnostics/Diagnostics/ViewModels/ControlDetailsViewModel.cs:
```diff
@@ -0,0 +1,185 @@
+                }
+                else
+                {
+                    return string.Compare(x.Name, y.Name);
```
Could use ordinal comparison instead of culture specific one.

MarchingCube commented on src/Avalonia.Diagnostics/Diagnostics/ViewModels/LogicalTreeNode.cs:
```diff
@@ -1,6 +1,3 @@
-// Copyright (c) The Avalonia Project. All rights reserved.
```
Any reason why half the files have license header and half do not?

grokys replied:
Simply, the ones where I modified an old file still have headers. I don't bother adding headers to files these days as it just feels like busywork and from what I've read legally it's meaningless.

grokys replied:
Maybe I should delete existing headers on modified files? Not sure.

MarchingCube replied:
TBH it would be good to be consistent at least. I have no problem with or without license info as long as we stick to one option.

MarchingCube commented on src/Avalonia.Diagnostics/Diagnostics/ViewModels/TreePageViewModel.cs:
```diff
@@ -66,7 +79,7 @@ public void SelectControl(IControl control)
                     control = control.GetVisualParent<IControl>();
                 }
-            }
+            }            
```
Nit: Whitespace.

MarchingCube commented on src/Avalonia.Diagnostics/Diagnostics/ViewModels/ViewModelBase.cs:
```diff
@@ -1,34 +1,42 @@
 namespace Avalonia.Diagnostics.ViewModels
 {
-    internal class ViewModelBase : INotifyPropertyChanged
+    public class ViewModelBase : INotifyPropertyChanged
```
Should be `internal`. We will end up with people using this as their base VM again.

grokys replied:
Arrgh? How did that come back?

MarchingCube commented on src/Avalonia.Diagnostics/Diagnostics/Views/EventsPageView.xaml.cs:
```diff
@@ -8,11 +8,11 @@
 namespace Avalonia.Diagnostics.Views
 {
-    public class EventsView : UserControl
+    public class EventsPageView : UserControl
```
`internal` ?

MarchingCube commented on src/Avalonia.Diagnostics/Diagnostics/Views/TreePageView.xaml.cs:
```diff
@@ -39,7 +36,7 @@ protected void AddAdorner(object sender, PointerEventArgs e)
                 {
                     Fill = new SolidColorBrush(0x80a0c5e8),
-                    [AdornerLayer.AdornedElementProperty] = node.Visual
+                    [AdornerLayer.AdornedElementProperty] = node.Visual,
```
Nit: Not needed comma.

grokys replied:
I actually prefer to leave trailing commas in initialization lists like this as it makes the diff cleaner when adding new entries.

MarchingCube reviewed (changes requested):
Glanced over, submitting my immediate feedback and I shall return later to evaluate it further :)

Functionality wise LGTM.

grokys committed:
Added some XML docs.

grokys committed:
Fixed nits.

MarchingCube reviewed (approved):
Looks like a great improvement over current one so I see no reason to not merge this in :)

grokys committed:
Merge branch 'master' into feature/new-devtools

kekekeks commented:
Note: we still need to make Roslyn dependency optional if we want devtools to be operational on iOS/WASM.

grokys commented:
TBH, I'm not sure how useful it would be there anyway, but yeah we can think about that when the time comes.

kekekeks commented:
For mobile the idea was to use our previewer protocol. You know, that one that can [show our controls in browser](https://github.com/AvaloniaUI/Avalonia/pull/3273).

Gillibald commented:
Currently the app itself shows the dev tools on demand. Why not make this a service you can connect to? So we could integrate this into the Visual Studio Extension or have it standalone.

Snoop for example could just inject itself into the target application and it just worked.
